src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -225,16 +226,122 @@ public override bool DirectoryExists(string fullPath)
+            {
+                yield return storageFiles[i].Path;
+            }
+        }
+
+        private async static Task<IReadOnlyList<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            string fullPath = Path.GetFullPath(path);
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(fullPath).TranslateWinRTTask(fullPath, isDirectory: true);
+            }
+            catch (FileNotFoundException e)
+            {
+                // GetFolderFromPathAsync throws FileNotFoundException but win32 throws a DirectoryNotFoundException for this case.
+                throw new DirectoryNotFoundException(e.Message, e.InnerException);
+            }
+
+            // Construct a query for the search.
+            QueryOptions query = new QueryOptions();
+
+            // Translate SearchOption into FolderDepth
+            query.FolderDepth = searchOption == SearchOption.AllDirectories ? FolderDepth.Deep : FolderDepth.Shallow;
+
+            // Construct an AQS filter
+            string normalizedSearchPattern = PathHelpers.NormalizeSearchPattern(searchPattern);
+            if (normalizedSearchPattern.Length == 0)
+            {
+                // An empty searchPattern will return no results and requires no AQS parsing.
+                return new IStorageItem[0];
+            }
+            else
+            {
+                // Parse the query as an ItemPathDisplay filter.
+                string searchPath = PathHelpers.GetFullSearchString(fullPath, normalizedSearchPattern);
+                string aqs = "System.ItemPathDisplay:~\"" + searchPath + "\"";
+                query.ApplicationSearchFilter = aqs;
+
+                // If the filtered path is deeper than the given user path, we need to get a new folder for it.
+                // This occurs when someone does something like Enumerate("C:\first\second\", "C:\first\second\third\*").
+                // When AllDirectories is set this isn't an issue, but for TopDirectoryOnly we have to do some special work
+                // to make sure something is actually returned when the searchPattern is a subdirectory of the path.
+                // To do this, we attempt to get a new StorageFolder for the subdirectory and return an empty enumerable
+                // if we can't.
+                string searchPatternDirName = Path.GetDirectoryName(normalizedSearchPattern);
+                string userPath = string.IsNullOrEmpty(searchPatternDirName) ? fullPath : Path.Combine(fullPath, searchPatternDirName);
+                if (userPath != folder.Path)
+                {
+                    try
+                    {
+                        folder = await StorageFolder.GetFolderFromPathAsync(userPath).TranslateWinRTTask(userPath, isDirectory: true);
```

Comment:
You're still doing an extra call to GetFolderFromPathAsync here.  You don't use folder anywhere between it's initial creation and here.  You could combine this call and the previous to avoid throw-away work.